python/cudf/cudf/core/resample.py

```diff
@@ -0,0 +1,361 @@
+
+    def _handle_frequency_grouper(self, by):
+        # if `by` is a time frequency grouper, we bin the key column
+        # using bin intervals specified by `by.freq`, then use *that*
+        # as the groupby key
+        import cudf._lib.labeling
+
+        freq = by.freq
+        label = by.label
+        closed = by.closed
+
+        # convert freq to a pd.DateOffset:
+        if isinstance(freq, str):
+            offset = pd.tseries.frequencies.to_offset(freq)
+        elif isinstance(freq, cudf.DateOffset):
+            offset = freq._maybe_as_fast_pandas_offset()
+        else:
+            if not isinstance(freq, pd.DateOffset):
+                raise TypeError(
+                    "Unsupported type for freq: {type(freq).__name__}"
+                )
+            offset = freq
+
+        if offset.freqstr == "M" or offset.freqstr.startswith("W-"):
+            label = "right" if label is None else label
+            closed = "right" if closed is None else closed
+        else:
+            label = "left" if label is None else label
+            closed = "left" if closed is None else closed
+
+        # determine the key column
+        if by.key is None and by.level is None:
+            # then assume that the key is the index of `self._obj`:
+            if not isinstance(self._obj.index, cudf.DatetimeIndex):
+                raise TypeError(
+                    f"Can only resample with a DatetimeIndex, "
+                    f"got {type(self._obj.index).__name__}"
+                )
+            self._handle_index(self._obj.index)
+        elif by.key:
+            self._handle_label(by.key)
+        elif by.level:
+            self._handle_level(by.level)
+
+        if not len(self._key_columns) == 1:
+            raise ValueError("Must resample on exactly one column")
+
+        key_column = self._key_columns[0]
+
+        if not isinstance(key_column, cudf.core.column.DatetimeColumn):
+            raise TypeError(
+                f"Can only resample on a DatetimeIndex or datetime column, "
+                f"got column of type {key_column.dtype}"
+            )
+
+        # get the start and end values that will be used to generate
+        # the bin labels
+        start, end = _get_timestamp_range_edges(
+            pd.Timestamp(key_column.min()),
+            pd.Timestamp(key_column.max()),
+            offset,
+            closed=closed,
+        )
+
+        # in some cases, an extra time stamp is required in order to
+        # bin all the values. It's OK if we generate more labels than
+        # we need, as we remove any unused labels below
+        end += offset
+
+        # generate the labels for binning the key column:
+        bin_labels = cudf.date_range(start=start, end=end, freq=freq,)
+
+        # We want the (resampled) column of timestamps in the result
+        # to have a resolution closest to the resampling
+        # frequency. For example, if resampling from '1T' to '1s', we
```

Comment:
since this is a place we diverge from pandas we should probably be loud about this in the docs, and note the limitations/future direction we settled on here